upstream-tag-sync/src/index.ts

```diff
@@ -49,66 +58,66 @@ async function retryWithBackoff<T>(
-                version: semver.valid(semver.clean(tag.name))
-            }))
-            .filter((tag): tag is { name: string, version: string } => tag.version !== null)
-            .sort((a, b) => semver.rcompare(a.version, b.version));
-
-        if (validTags.length === 0) {
-            warning('No semver-compliant tags found in repository');
-            return null;
-        }
-
-        const latestTag = validTags[0];
-        info(`Found latest tag: ${latestTag.name} (${latestTag.version})`);
-        return latestTag.name;
+  try {
+    const iterator = octokit.paginate.iterator(octokit.rest.repos.listTags, {
+      owner,
+      repo,
+      per_page: 100,
+    });
 
-    } catch (err) {
-        const error = err as Error;
-        if (attempt >= maxAttempts) {
-            throw new Error(`Failed to fetch tags after ${maxAttempts} attempts: ${error.message}`);
-        }
+    const tags: { name: string }[] = [];
+    for await (const { data: pageTags } of iterator) {
+      tags.push(...pageTags);
+      info(`Fetched ` + tags.length.toString() + ` tags so far...`);
 
-        // Calculate delay with exponential backoff and jitter
-        const delay = Math.min(baseDelay * Math.pow(2, attempt - 1) + Math.random() * 1000, 60000); // Cap at 60 seconds
-        info(`Error fetching tags (attempt ${attempt}/${maxAttempts}), waiting ${Math.round(delay / 1000)}s before retry...`);
-        await new Promise(resolve => setTimeout(resolve, delay));
+      // Add a small delay between pages to be nice to the API
+      await new Promise(resolve => setTimeout(resolve, 1000));
+    }
+
+    if (tags.length === 0) {
+      info('No tags found in repository');
+      return null;
+    }
 
-        return getLatestTag(octokit, owner, repo, attempt + 1, maxAttempts, baseDelay);
+    const validTags = tags
+      .map(tag => ({
+        name: tag.name,
+        version: semver.valid(semver.clean(tag.name))
+      }))
+      .filter((tag): tag is { name: string, version: string } => tag.version !== null)
+      .sort((a, b) => semver.rcompare(a.version, b.version));
+
+    if (validTags.length === 0) {
+      warning('No semver-compliant tags found in repository');
+      return null;
     }
+
+    const latestTag = validTags[0];
+    info(`Found latest tag: ${latestTag.name} (${latestTag.version})`);
+    return latestTag.name;
+
+  } catch (err) {
+    const error = err as Error;
+    if (attempt >= maxAttempts) {
+      throw new Error(`Failed to fetch tags after ` + maxAttempts.toString() + ` attempts: ${error.message}`);
+    }
+
+    // Calculate delay with exponential backoff and jitter
+    const delay = Math.min(baseDelay * Math.pow(2, attempt - 1) + Math.random() * 1000, 60000); // Cap at 60 seconds
+    info(`Error fetching tags (attempt ` + attempt.toString() + `/` + maxAttempts.toString() + `), waiting ` + Math.round(delay / 1000).toString() + `s before retry...`);
+    await new Promise(resolve => setTimeout(resolve, delay));
```

Comment:
Again, I'm confused why we don't use `retryWithBackoff()` here

Reply:
Sorry for the code above with trying to retry an `await iterator.next()`. I'm actually pretty certain this whole idea of retrying on an iterator is wrong and might not be possible at all. I changed to use the simple version _WITH_ using `retryWithBackoff()`.